shared/types/mempool.go

```diff
@@ -0,0 +1,142 @@
+package types
+
+import (
+	"bytes"
+	"container/list"
+	"encoding/hex"
+	"github.com/pokt-network/pocket/shared/crypto"
+	"sync"
+)
+
+type Mempool interface {
+	Contains(hash string) bool
+	AddTransaction(tx []byte) Error
+	DeleteTransaction(tx []byte) Error
+
+	Flush()
+	Size() int
+	TxsBytes() int
+	PopTransaction() (tx []byte, sizeInBytes int, err Error)
+}
+
+var _ Mempool = &FIFOMempool{}
+
+type FIFOMempool struct {
+	l                    sync.RWMutex
+	hashMap              map[string]struct{}
+	pool                 *list.List
+	size                 int
+	transactionBytes     int
+	maxTransactionsBytes int
+	maxTransactions      int
+}
+
+func NewMempool(maxTransactionBytes int, maxTransactions int) Mempool {
+	return &FIFOMempool{
+		l:                    sync.RWMutex{},
+		hashMap:              make(map[string]struct{}),
+		pool:                 list.New(),
+		size:                 0,
+		transactionBytes:     0,
+		maxTransactionsBytes: maxTransactionBytes,
+		maxTransactions:      maxTransactions,
+	}
+}
+
+func (f *FIFOMempool) AddTransaction(tx []byte) Error {
+	f.l.Lock()
+	defer f.l.Unlock()
+	hash := crypto.SHA3Hash(tx)
+	hashString := hex.EncodeToString(hash)
+	if _, ok := f.hashMap[hashString]; ok {
+		return ErrDuplicateTransaction()
+	}
+	f.pool.PushBack(tx)
+	f.hashMap[hashString] = struct{}{}
+	f.size++
+	f.transactionBytes += len(tx)
+	for f.size >= f.maxTransactions || f.transactionBytes >= f.maxTransactionsBytes {
```

Comment:
Maybe we should check this first before pushing into the pool so we don't need to do an unnecessary pop?